prairielib/lib/migrations.js

```diff
@@ -34,136 +35,272 @@ module.exports.init = function (dir, proj, callback) {
+
+        // Validation: if we no longer have any indexes in the migration names,
+        // ensure that the user has deployed an earlier version of the code that
+        // already synced the indexes to the migrations table.
+        if (!migrationFiles.some((m) => m.index !== null)) {
+          const migrationsMissingTimestamps = allMigrations.filter((m) => !m.timestamp);
+          if (migrationsMissingTimestamps.length > 0) {
+            const missing = migrationsMissingTimestamps.map((m) => m.filename).join(', ');
+            throw new Error(`The following migrations are missing timestamps: ${missing}`);
           }
+        }
 
-          files = files.filter((file) => file.index > last_migration);
-          callback(null, files);
-        });
-      },
-      (files, callback) => {
-        async.eachSeries(
-          files,
-          (file, callback) => {
-            if (noExistingMigrations) {
-              // if we are running all the migrations then log at a lower level
-              logger.verbose('Running migration ' + file.filename);
-            } else {
-              logger.info('Running migration ' + file.filename);
-            }
-            async.waterfall(
-              [
-                (callback) => {
-                  fs.readFile(path.join(migrationDir, file.filename), 'utf8', (err, sql) => {
-                    if (ERR(err, callback)) return;
-                    callback(null, sql);
-                  });
-                },
-                (sql, callback) => {
-                  // Perform the migration
-                  sqldb.query(sql, [], (err, _result) => {
-                    if (err) error.addData(err, { sqlFile: file.filename });
-                    if (ERR(err, callback)) return;
-                    callback(null);
-                  });
-                },
-                (callback) => {
-                  // Record the migration
-                  const params = {
-                    filename: file.filename,
-                    index: file.index,
-                    project,
-                  };
-                  sqldb.query(sql.insert_migration, params, (err, _result) => {
-                    if (ERR(err, callback)) return;
-                    callback(null);
-                  });
-                },
-              ],
-              (err) => {
-                if (ERR(err, callback)) return;
-                callback(null);
-              }
-            );
-          },
-          (err) => {
-            if (ERR(err, callback)) return;
-            callback(null);
+        // Second pass: reconcile the timestamps with the list of migrations in the database.
+        // This should only matter a single time (the first time this code is deployed after
+        // adding timestamps to all migrations).
+        for (const { filename, timestamp, index } of migrationFiles) {
+          if (!timestamp) {
+            continue;
```

Comment:
I originally didn't do that because I wanted to make sure this wouldn't block PT migrations (i.e. I wanted to be able to consume a new PL version without immediately cutting over to new migration filenames). Thinking about that some more though, we should probably just rip the bandaid off and enforce that.

Reply:
I made timestamps mandatory in 3a56633.

Reply:
Oh right, good point. But I agree with also converting PT immediately.